### ISSUE TITLE
Would it make sense to use the embeddings for transfer learning / binary classification?

### ISSUE THREAD

**Issue by timsuchanek** [UKPLab/sentence-transformers#209]
I'm tackling a problem, where I want to have individual text classification models for users based on their preferences. It's, therefore, a binary classification task, with the label either being like or dislike.

While [huggingface/transfomers](https://github.com/huggingface/transformers) already is great for classification tasks, in my understanding, they add a dropout and linear layer on top of the BERT models and perform the feed-forward during training through the whole network https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_albert.py#L694

If you're building a platform that serves many users, where you want to have an individually trained model per user, this would require a lot of computing.

Instead, I could imagine just getting the embeddings of the texts I want to classify with `sentence-transformers` and using K-nearest neighbors to perform text classification. I'm aware, that the results would probably not be as great as using the training of the transformers library, but would they at least make some sense in your opinion?

This would be a very compute effective way to scale text classification across many users (if it works), even if it's not having the best results.

Another approach would be, to add a minimal "head" on top of the embedding calculation, e.g. just a few hidden layers to train binary classification, similar to https://arxiv.org/abs/1801.06146 or https://github.com/hanxiao/bert-as-service/blob/master/example/example5.py#L66

I hope these thoughts somehow make sense.

Thanks!

**Comment by nreimers** [comment]
One of my students (@aaronsom) is I tackling a similar issue. There, we have also a large set of individual models (10k individual models, each trained on a different dataset), that perform a binary classification on a given input sentence. 

He passes the input sentence through a fixed transformer network. The output vector (he takes the representation of the CLS token) is than passed through these 10k individual binary classification models to derive a final label which of the 10k models are suitable. Each binary classification model is just a simple softmax classifier, i.e. a 768 dimensional weight vector that is multiplied and summed with the output sentence embedding. In order to make this efficient, he implemented this with a convolutional layer, so that it can be computed within a few milliseconds on a GPU.

Maybe @aaronsom can help if you need further information.

Best
Nils Reimers

**Comment by Aaronsom** [comment]
As Nils already wrote, I have a somewhat similar problem with many independent classifiers using the same underlying transformer model.

You can try to use a pretrained model (when using the CLS token, you don't neccessarily need a sentence transformer, a "basic" huggingface model is enough) or you can finetune on a similar auxiliary task so the representation is more relevant for your classifiers. 

Using convolutions is a neat trick if you want to have the results of all classifiers at the same time in an efficient manner (see e.g. https://sebastianraschka.com/faq/docs/fc-to-conv.html). 
However, if you just want to know if one user would like one text at a time, you can ignore this and use a normal linear layer with sigmoid.

Also, you should still try the knn idea, that might work better than expected in my experience.

**Comment by timsuchanek** [comment]
Perfect, thanks a lot for the answers!
I'll try out both and report back here how it goes :)
